### PR TITLE
routing: Transparently convert a MultiDict containing parameters.

### PR DESCRIPTION
**Pull request by houseofsuns**
Sorry to nag once more w.r.t. the multiple values for URL building, but I think I found a way to make everybody happy. The following should allow passing in MultiDicts and receiving the desired behaviour without getting any regressions.
This should be ready to merge (as in there is test and documentation).

This allows to pass a MultiDict and get the intuitively expected result
(that is multiple parameters, where multiple values are present). This
allows to recreate a URL as follows:
- create a URL with multiple values for a parameter
- use Map.match() to extract them into a MultiDict
- recreate the URL with Map.build() handing in the MultiDict

The last step would fail before this commit (in that it would reproduce only
one of the values for each parameter). Also this avoids the collateral
damage introduced by the last attempt at allowing MultiDicts.

This zaps one of the test cases in test_basic_building to reflect the
changed behaviour.

### CODE REVIEW

**Comment by veelai** [comment]
Hm, I messed up the layout above ...

**Comment by untitaker** [comment]
You can pass in multiple query params by providing a list of values as value. Isn't that good enough?

**Comment by veelai** [comment]
I have no strong feelings here, but I found the point that build() is then inverse to match() a big plus. For me this is more an enhancement/sugar coating so that everything works as I would intuitively expect it.

Also if not this change, I will have to modify some of my code to use lists of values, so I thought I might try here first, for the benefit of the general public. :)

**Comment by veelai** [comment]
I just saw the failure and fixed the syntax to be compatible to 2.6

File: werkzeug/routing.py [hunk 1]
```diff
@@ -1742,10 +1748,14 @@ def build(self, endpoint, values=None, method=None, force_external=False,
         self.map.update()
         if values:
             if isinstance(values, MultiDict):
-                valueiter = iteritems(values, multi=True)
+                temp = dict(
+                    (k, (values.getlist(k)
+                         if len(values.getlist(k)) > 1
```
Comment (untitaker) [inline werkzeug/routing.py]: This is really slow, it calls getlist twice for no reason. I'll try to write some benchmarks, but the current code isn't acceptable (though I apprechiate it that there are no perf regressions for the non-multidict path)

Comment (veelai) [inline werkzeug/routing.py]: What about the following? It is imho a tad uglier, but should be pretty optimal in performance (in that there is nothing done which could be spared and keep the functionality the same).

    temp = {}
    for key, list_value in values.lists():
          if len(list_value) == 1:
              temp[key] = list_value[0]
          else:
                 temp[key] = list_value

Comment (untitaker) [inline werkzeug/routing.py]: This looks good but please use `iterlists` from compat.

**Comment by veelai** [comment]
Tentative bump. What about this, it looks like I'll actually not need this anymore, but I think it would be a sensible change anyway. :)

**Comment by davidism** [comment]
Issue history so I can keep this straight:

* #658 was the initial request to support multiple values when building.
* #660 converted the values to a `MultiDict` unconditionally.
* #667 fixed an issue #660 caused when passing dicts.
* #709 reported an issue with being able to indicate a single, iterable value vs multiple values.
* #710 reverted #660 to return to the previous behavior.

**Comment by davidism** [comment]
Continued in #1310

**Pull request closed without merging**